File: CSharp/Math/Ray.cs
using System;
using System.Collections.Generic;
using System.Text;

namespace CSharp
{
    public class Ray
    {
        public Ray(Point origin, Vector direction)
        {
            Origin = origin;
            Direction = direction;
        }

        public Point Origin { get; }
        public Vector Direction { get; }

        public Point Position(double t)
        {
            return new Point(Origin + Direction * t);
        }

        public Ray Transform(Matrix m)
        {
            var origin = m * Origin;
            var direction = m * Direction;

            return new Ray(origin, direction);
        }

        public override bool Equals(object obj)
        {
            var other = obj as Ray;
            return ((other != null) &&
                    Origin.Equals(other.Origin) &&
                    Direction.Equals(other.Direction));
        }

        public override int GetHashCode()
        {
            return HashCode.Combine(Origin, Direction);
        }

        public override string ToString()
        {
            return $"Ray {Origin} {Direction}";
        }
    }
}

File: Java/src/main/java/Math/Ray.java
package RTC.Math;

import java.util.Objects;

public class Ray {
    public static int RayId = 0;

    private int id;
    private Point origin;
    private Vector direction;

    public Ray(Point Origin, Vector Direction) {
        id = RayId++;

        origin = Origin;
        direction = Direction;
    }

    public Point getOrigin() {
        return origin;
    }

    public Vector getDirection() {
        return direction;
    }

    public Point Position(double t) {
        Tuple d = direction.mul(t);
        d = origin.add(d);

        return new Point(d);
    }

    public Ray transform(Matrix m) {
        Tuple o = m.mul(origin);
        Tuple d = m.mul(direction);

        return new Ray(new Point(o), new Vector(d));
    }

    @Override
    public boolean equals(Object o)
    {
        if (o == this) {
            return true;
        }

        if ((o == null) || !(o instanceof Ray)) {
            return false;
        }

        Ray r = (Ray) o;
        return origin.equals(r.getOrigin()) &&
               direction.equals(r.getDirection());
    }

    @Override
    public int hashCode() {
        return Objects.hash(id, origin.hashCode(), direction.hashCode());
    }

    @Override
    public String toString() {
        return String.format("Ray [%s, %s] Id: %d", origin.toString(), direction.toString(), id);
    }
}
